### ISSUE TITLE
Some replays throw errors

### ISSUE THREAD

**Issue by Buszmen** [tfausak/rattletrap#51]
I'm using rattletrap for Discord bot and after testing I've got some replays that throw errors. 
All of them work on Rocket League. 

Tested on Rattletrap 3.0.0 (Windows / Linux)

https://cdn.discordapp.com/attachments/279639830777036801/375330181293932544/2CFE577044E651D3FA9DBF83ECA8BCC3.replay
https://cdn.discordapp.com/attachments/279639830777036801/375328580986470400/6E905EE8419F3DE8BC7A18A55811B93F.replay
https://cdn.discordapp.com/attachments/279639830777036801/375327305213542400/7C73CCCC4BCB9FBE8EA11EAF79888987.replay
https://cdn.discordapp.com/attachments/279639830777036801/375328845659766784/08A88B8C43418CB4E3D0988BEC5EB9C0.replay
https://cdn.discordapp.com/attachments/279639830777036801/375330054793723908/DE56B1B146687EA2E62E31A37A08C8EB.replay
https://cdn.discordapp.com/attachments/279639830777036801/375328363541168128/F95FA7A345970D2F69CFD6A1DE04CAD7.replay

**Comment by tfausak** [comment]
Bummer! I'll take a look later tonight. If you can, posting the actual errors that Rattletrap spits out would be helpful. If not, I'll do that tonight.

**Comment by Buszmen** [comment]
Sure thing

2CFE577044E651D3FA9DBF83ECA8BCC3.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 158521: not enough bytes
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get`

6E905EE8419F3DE8BC7A18A55811B93F.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 769: unknown object name Just
 (Text {textValue = "TAGame.Default__CameraSettingsActor_TA"}) for ID Word32 {word32Value = 146}
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get`

7C73CCCC4BCB9FBE8EA11EAF79888987.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 2419: unknown object name Jus
t (Text {textValue = "Engine.Pawn:RootMotionInterpRate"}) for ID Word32 {word32Value = 142}
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get`

08A88B8C43418CB4E3D0988BEC5EB9C0.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 1799: unknown object name Just (Text {textValue = "TAGame.PRI_TA:ClientLoadoutsOnline"}) for ID Word32 {word32Value = 141}
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get`

DE56B1B146687EA2E62E31A37A08C8EB.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 1499: unknown object name Just (Text {textValue = "Engine.PlayerReplicationInfo:bIsSpectator"}) for ID Word32 {word32Value = 141}
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5
.1:Data.Binary.Get`

F95FA7A345970D2F69CFD6A1DE04CAD7.replay
`rattletrap.exe: Data.Binary.Get.runGet at position 2112: missing object name for ID Word32 {word32Value = 33359465}
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get`

**Comment by tfausak** [comment]
Thanks! The first and last ones are baffling, but the rest should be easy to fix.

**Comment by tfausak** [comment]
Slightly more readable error messages:

```
08A88B8C43418CB4E3D0988BEC5EB9C0
rattletrap: Data.Binary.Get.runGet at position 1799: unknown object name Just (Text {textValue = "TAGame.PRI_TA:ClientLoadoutsOnline"}) for ID Word32 {word32Value = 141}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get

2CFE577044E651D3FA9DBF83ECA8BCC3
rattletrap: Data.Binary.Get.runGet at position 158521: not enough bytes
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get

6E905EE8419F3DE8BC7A18A55811B93F
rattletrap: Data.Binary.Get.runGet at position 769: unknown object name Just (Text {textValue = "TAGame.Default__CameraSettingsActor_TA"}) for ID Word32 {word32Value = 146}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get

7C73CCCC4BCB9FBE8EA11EAF79888987
rattletrap: Data.Binary.Get.runGet at position 2419: unknown object name Just (Text {textValue = "Engine.Pawn:RootMotionInterpRate"}) for ID Word32 {word32Value = 142}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get

DE56B1B146687EA2E62E31A37A08C8EB
rattletrap: Data.Binary.Get.runGet at position 1499: unknown object name Just (Text {textValue = "Engine.PlayerReplicationInfo:bIsSpectator"}) for ID Word32 {word32Value = 141}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get

F95FA7A345970D2F69CFD6A1DE04CAD7
rattletrap: Data.Binary.Get.runGet at position 2112: missing object name for ID Word32 {word32Value = 33359465}
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:342:5 in binary-0.8.3.0:Data.Binary.Get
```

And all the replays zipped up for posterity: 
[rattletrap-issue-51-replays.zip](https://github.com/tfausak/rattletrap/files/1448731/rattletrap-issue-51-replays.zip)

**Comment by tfausak** [comment]
This is going to take me a while to fix. The way the code is structured now, product attributes are off in their own little world. (Check out [`getProductAttribute`](https://github.com/tfausak/rattletrap/blob/3.1.0/library/Rattletrap/Attribute/ProductAttribute.hs#L29) for how that works.) Based on the error messages, it seems like product attributes are really just slightly different from normal attributes. Combining those two types and their code paths might fix most of the broken replays.

**Comment by Buszmen** [comment]
I'm not a Haskell programmer so it may be wrong suggestion, but maybe there is a way to make a workaround (at least if the actual fix is so hard to achieve). We have something like graceful degradation. In my case I don't use all the data from generated JSON. I also don't know how those errors would impact the result, but if I can get partial data instead of no data it may be enough for me or some others who use this tool.

**Comment by tfausak** [comment]
Currently using Rattletrap as a command line application, there's no way to output only some of the data. It would be nice to have (#55), but I'm not sure what it would look like. 

It might make sense to dump the output up to the failure when something like this happens. I could see Rattletrap outputting partial JSON to STDOUT, the error information to STDERR, and returning a non-zero status code.

**Comment by danielsamuels** [comment]
I have ~700 failing replays on RLR at the moment, I can provide a list of links if that would be useful.

**Comment by tfausak** [comment]
Nah, I should be able to figure it out with the replays I've already got. It'll just take some time.

**Comment by tfausak** [comment]
I haven't forgotten about this. I would like Rattletrap to work with the latest replays, but it's a pretty low priority for me right now.

**Comment by enzanki-ars** [comment]
I am seeing similar errors on 3.1.0 with a bunch of replay files.

[00120002001C000C000401B0C7623773.zip](https://github.com/tfausak/rattletrap/files/1549943/00120002001C000C000401B0C7623773.zip)

```
rattletrap-3.1.0-windows.exe: Data.Binary.Get.runGet at position 27416: not enough bytes
CallStack (from HasCallStack):
  error, called at libraries\binary\src\Data\Binary\Get.hs:351:5 in binary-0.8.5.1:Data.Binary.Get
```

(Sidenote: I am using Windows, but these files were sent to me from a Mac user.  Highly doubt that will cause any issues, but thought I'd mention just incase.)

**Comment by tfausak** [comment]
Yeah the 3.1.0 release didn’t include a fix for this. 

For what it’s worth, I haven’t seen a difference between Windows, macOS, and Linux.

**Comment by enzanki-ars** [comment]
Really wish I could help out here, but I don't understand Haskell nor do I have enough knowledge of the binary format.  I do partially understand the resulting JSON though!  

Thank you for creating an amazing utility!

**Comment by tfausak** [comment]
Thanks! 

As I said before, it would be nice if Rattletrap worked with the latest replays. I thought I had fixed it earlier (#48), but that turned out to be wrong. 

I've been working on parsing Rocket League replays for like two years now. It's getting kind of tedious to keep up with the changes to the file format and I'm losing steam. 

There are other projects that parse replays: https://github.com/rocket-league-replays/rocket-league-replays/wiki/Rocket-League-Replay-Parsers

@jjbott generally seems to know his way around these files better than I do. It might be worth trying out his parser until I find the time/motivation to fix Rattletrap.

**Comment by enzanki-ars** [comment]
Thanks for the tip! I'll be sure to try it out!

**Comment by jjbott** [comment]
Don't give up! It's no fun without competition! :)

[My parser](https://github.com/jjbott/RocketLeagueReplayParser) does seem to work fine with all of the above if you need sample code. Or, if anyone wants to try mine and needs help, feel free to create an issue and I'll check it out. My parser is barely maintained and the code is getting pretty scary, but I do try to make sure it can parse everything.

**Comment by enzanki-ars** [comment]
@jjbott Thanks!  I am going through and updating my project to use your parser, and it is pretty nice.  I like how you have flattened the data down too!  Makes it a lot easier to parse.  My only concern is users that want to use my project on Linux/Mac.  I do know C# (it's been a while though...), so I may help out a bit if I see something in the output that looks weird, but so far I am loving it and haven't had any issues.

(In case you are curious what project I am referencing: https://github.com/enzanki-ars/rocket-league-minimap-generator)  Demo video will be uploaded soon, but the parsing of the data for newer versions got in my way.

**Comment by tfausak** [comment]
I'm having a hard time understanding why @jjbott's parser works with these new replays but mine doesn't. I know Haskell isn't the easiest for everyone to read, but it looks like we're both doing more or less the same thing:

- https://github.com/jjbott/RocketLeagueReplayParser/blob/884a4fc/RocketLeagueReplayParser/NetworkStream/ProductAttribute.cs#L27
- https://github.com/tfausak/rattletrap/blob/e88f4be/library/Rattletrap/Attribute/Product.hs#L35

My parser must be doing something wrong before it even reaches the product attributes.

**Comment by tfausak** [comment]
Aha! These lines might have something to do with it: https://github.com/jjbott/RocketLeagueReplayParser/blob/884a4fc/RocketLeagueReplayParser/NetworkStream/ClientLoadoutOnline.cs#L28-L31

My code doesn't have that logic. I'll add it in and see if that helps.

**Comment by jjbott** [comment]
Honestly, I think that specific logic is nonsense.  At one point ages ago I was curious if anyone was copying my logic as-is, so I put some trash in there. I don't know if anyone copied that logic because I didn't care enough to check, so it was a huge waste of time. :)

**Comment by tfausak** [comment]
I just added that logic and it didn't help at all, so you might be right there 😄 

This could end up being like #48 where some far-off thing is subtly wrong and it's only showing up here. Have I mentioned that I'm not a fan of formats that aren't byte aligned?

**Comment by tnunamak** [comment]
@tfausak your parser is the basis for [a tool](https://github.com/tnunamak/rocket-league-utils) that helps the [ORSA Dynasty Cup](https://www.reddit.com/r/RocketLeague/comments/73ook2/orsa_leagues_are_back_better_and_bigger_than_ever/?st=jb3uup49&sh=f8630b92) league of ~150 players (and growing) collect stats throughout each two month season. Organizers have a manual fallback, but they love to ask me when they'll be able to parse stats automatically again. I tend to derive the most motivation for my OSS projects when I know that I have users who care... so I wanted to let you know that you have users who care 😄. Feel free to stop by the [discord server](https://www.discord.me/dynastycup) if you ever want to say hello (or play!). Thanks for providing this tool to the community!

**Comment by tfausak** [comment]
Thanks for the kind words @tnunamak! That means a lot 😍 

I was trying to get @jjbott's parser set up on my machine so I could compare the non-frame output, but I don't know the first thing about .NET development. How can I get a binary installed for parsing replays? I'm on Windows and installed NuGet and did `nuget install RocketLeagueReplayParser`, but that only gave me a `.nupkg` and a `.dll`. I don't see a way to download binaries on either NuGet, nor do I see any recent binaries on GitHub.

**Comment by jjbott** [comment]
Yeah, mine is more tailored to be used as a library. There is a console front end, but no one uses it so I forget it exists. :) But I've updated it, maybe that'll help!

I also saved the output from a few and attached them here

[ParsedReplays.zip](https://github.com/tfausak/rattletrap/files/1554069/ParsedReplays.zip)

**Comment by tfausak** [comment]
Thanks! Looking over the non-frame stuff for the `2CFE...` replay, everything seems to match. Unfortunately your parser doesn't spit out all the boring details (like the class property map), so all I can say is that the header information looks fine. 

By adding some debugging output to my parser and comparing it to your parser's output, I can see that mine is failing to parse this frame: https://gist.github.com/tfausak/ad6125d8eec91a7d2df06bb3e5c070ff

Nothing in that frame looks troubling. In fact, it's a small frame. The only interesting attribute replicated in it is a reservation. 

Anyway, thanks again for the JSON. I'll take a closer look tomorrow.

**Comment by tfausak** [comment]
Aha! I think I'm finally making some headway here. Somehow I missed that PlayStation IDs got another 8 bytes in the recent patch. My code is definitely not doing that. All this replicated product attribute stuff was (maybe) a red herring. https://github.com/jjbott/RocketLeagueReplayParser/blob/884a4fc/RocketLeagueReplayParser/NetworkStream/UniqueId.cs#L46-L53

**Comment by tfausak** [comment]
I fixed the parsing of PlayStation IDs in new replays and released that as [3.1.1](https://github.com/tfausak/rattletrap/releases/tag/3.1.1). Executables should be available as soon as the builds finish. 

Rattletrap still can't parse all the replays listed in this issue, but it should be a little better.

**Comment by tfausak** [comment]
Wow, uh, this is embarrassing 😑 My product attribute parser was off by a *single bit*. Sorry for taking so long to fix this! [Rattletrap version 3.1.2](https://github.com/tfausak/rattletrap/releases/tag/3.1.2) can parse all the replays in this issue. The binaries should be available soon.
